grpc-sys/build.rs

```diff
@@ -40,156 +38,92 @@ mod imp {
 
-    const ZLIB_VERSION: &'static str = "1.2.8";
+    const ZLIB_VERSION: &'static str = "v1.2.8";
     const BORINGSSL_COMMIT_HASH: &'static str = "78684e5b222645828ca302e56b40b9daff2b2d27";
     const CARES_TAG: &'static str = "cares-1_12_0";
 
-    struct GithubRepoFetcher {
-        url: String,
-        base_name: String,
-        to_dir: String,
-    }
-
-    impl GithubRepoFetcher {
-        fn new_with_hash(account: &str,
-                         repo: &str,
-                         commit_hash: &str,
-                         to_dir: &str)
-                         -> GithubRepoFetcher {
-            GithubRepoFetcher {
-                url: format!("https://github.com/{}/{}/archive/{}.tar.gz",
-                             account,
-                             repo,
-                             commit_hash),
-                base_name: format!("{}-{}", repo, commit_hash),
-                to_dir: to_dir.to_owned(),
+    fn execute(cmd: &mut Command) {
+        match cmd.status() {
+            Err(e) => panic!("failed to execute {:?}: {}", cmd, e),
+            Ok(status) => {
+                if !status.success() {
+                    panic!("command {:?} exit with {}", cmd, status);
+                }
             }
         }
+    }
 
-        fn new_with_tag(account: &str, repo: &str, tag: &str, to_dir: &str) -> GithubRepoFetcher {
-            GithubRepoFetcher {
-                url: format!("https://github.com/{}/{}/archive/{}.tar.gz",
-                             account,
-                             repo,
-                             tag),
-                base_name: format!("{}-{}", repo, tag),
-                to_dir: to_dir.to_owned(),
-            }
+    fn fetch_and_extract(account: &str, repo: &str, arch_tag: &str, to_dir: &str) {
+        let url = format!("https://github.com/{}/{}/archive/{}.tar.gz",
+                          account,
+                          repo,
+                          arch_tag);
+        let out_dir = env::var("OUT_DIR").unwrap();
+        let tgz_file_name = format!("{}-{}.tar.gz", repo, arch_tag);
+        let cmds = vec![
+            vec!["wget", "-q", "-O", &tgz_file_name, &url],
+            vec!["rm", "-rf", &to_dir],
+            vec!["mkdir", "-p", &to_dir],
+            vec!["tar", "zxf", &tgz_file_name, "-C", to_dir, "--strip-components", "1"],
+        ];
+        for cmd in cmds {
+            execute(Command::new(cmd[0]).args(&cmd[1..]).current_dir(&out_dir));
         }
+    }
 
-        fn new_with_semver(account: &str, repo: &str, ver: &str, to_dir: &str) -> GithubRepoFetcher {
-            GithubRepoFetcher {
-                url: format!("https://github.com/{}/{}/archive/v{}.tar.gz",
-                             account,
-                             repo,
-                             ver),
-                base_name: format!("{}-{}", repo, ver),
-                to_dir: to_dir.to_owned(),
-            }
-        }
+    fn inflate_grpc() {
```

Comment:
The function create the grpc directory and download all necessary dependencies for it from scratch.